robot/src/main/java/us/ilite/robot/modules/Elevator.java

```diff
@@ -0,0 +1,267 @@
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.team254.lib.util.Units;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.team254.lib.drivers.talon.TalonSRXFactory;
+
+import us.ilite.common.config.SystemSettings;
+import us.ilite.lib.drivers.SparkMaxFactory;
+
+import edu.wpi.first.wpilibj.Talon;
+import edu.wpi.first.wpilibj.Timer;
+// import sun.security.krb5.internal.rcache.DflCache;
+import us.ilite.robot.Data;
+import us.ilite.common.lib.control.PIDController;
+import us.ilite.common.lib.control.PIDGains;
+import com.team254.lib.util.Util;
+
+
+public class Elevator extends Module {
+
+
+    private boolean mAtBottom = true;
+    private boolean mAtTop = false;
+    private int mCurrentEncoderTicks;
+    private double mDesiredPower = 0;
+    private boolean mAtDesiredPosition;
+    private boolean mDesiredDirectionUp, mDesiredDirectionDown;
+    private boolean mDesiredPositionAboveInitial = false;
+    private double mMaxPower = 1.0d;
+    private double mMinPower = -1.0d;
+    private double mHoldVoltage = 0.5; //NOT final
+    private boolean mSettingPosition = false;
+    private int kCansparkId = 0; //TODO change ID
+    private int kTalonId = 0; //TODO change ID
+
+    //TODO need to figure out these encoder thresholds
+    private int mDownEncoderThreshold = SystemSettings.kLowerElevatorEncoderThreshold;
+    private int mUpEncoderThreshold = SystemSettings.kLowerElevatorEncoderThreshold;
+
+    EElevatorPosition mCurrentPosition = EElevatorPosition.BOTTOM;
+
+    EElevatorState mCurrentState = EElevatorState.NORMAL;
+    EElevatorPosition mDesiredPosition;
+
+    CANSparkMax mMasterElevator;
+    TalonSRX mEncoder = TalonSRXFactory.createDefaultTalon(0);
+
+    public Elevator(Data pData) {
+        //Create default NEO and set the ramp rate
+        mMasterElevator = SparkMaxFactory.createDefaultSparkMax(kCansparkId, MotorType.kBrushless);
+        mMasterElevator.setIdleMode(IdleMode.kBrake);
+        mMasterElevator.setRampRate(SystemSettings.kELevatorControlLoopPeriod);
+
+        //We start at the bottom
+        mAtBottom = true;
+        mAtTop = false;
+        mDesiredDirectionUp = true;
+
+        //Make sure the elevator is stopped upon initialization
+        mDesiredPosition = EElevatorPosition.BOTTOM;
+        mCurrentState = EElevatorState.STOP;
+
+        mCurrentEncoderTicks = 0;
+    }
+
+    public void shutdown(double pNow) {
+
+    }
+
+    public void modeInit(double pNow) {
+
+    }
+
+    public void periodicInput(double pNow) {
+
+    }
+
+    public void update(double pNow) {
+
+        //Get encoder value from secondary Talon
+        mCurrentEncoderTicks = getEnocderPosition();
+        mDesiredDirectionUp = (mDesiredPower > 0);
+
+        updateElevator(pNow);
+        mDesiredPower = determineActualPower(mCurrentState);
+
+        mMasterElevator.set(Util.limit(mDesiredPower, mMinPower, mMaxPower));
```

Comment:
Also if my last comment is wrong, the limit could potentially change the value of the desired power. Also recommend moving the limit inside of the determineActualPower method so that the actual power always returns a value within the acceptable range.